mqtt/policy/src/core/mod.rs

```diff
@@ -843,4 +833,98 @@ pub(crate) mod tests {
             policy.starts_with(input)
         }
     }
+
+    #[cfg(feature = "proptest")]
+    mod proptests {
+        use crate::{Decision, Effect, PolicyBuilder, PolicyDefinition, Request, Statement};
+        use proptest::{collection::vec, prelude::*};
+
+        proptest! {
+            /// The goal of this test is to verify the following scenarios:
+            /// - PolicyBuilder does not crash.
+            /// - All combinations of identity/operation/resource in a statement in the definition
+            ///   should produce expected result.
+            ///   Since some statements can be overridden by the previous ones,
+            ///   we can only safely verify the very first statement.
+            #[test]
+            fn policy_engine_proptest(definition in arb_policy_definition()){
+                use itertools::Itertools;
+
+                // take very first statement, which should have top priority.
+                let statement = &definition.statements()[0];
+                let expected = match statement.effect() {
+                    Effect::Allow => Decision::Allowed,
+                    Effect::Deny => Decision::Denied,
+                };
+
+                // collect all combos of identity/operation/resource
+                // in the statement.
+                let requests = statement.identities()
+                    .iter()
+                    .cartesian_product(statement.operations())
+                    .cartesian_product(statement.resources())
+                    .map(|item| Request::new(item.0.0, item.0.1, item.1)
+                                .expect("unable to create a request")
+                        ).collect::<Vec<_>>();
```

Comment:
```suggestion
                        );
```
I believe you don't need to allocate a vec only to iterate it later in for loop

Reply:
I needed to materialize the collection, b/c I can't continue to use borrowed statement after I create a policy from it. Policy consumes the definition.

Reply:
ah, that's right!